File: frontend/src/types/production.ts
/**
 * Greena — Production Readiness types (Module 11).
 * Mirrors backend/app/schemas/production.py.
 */

// ── Version / release ────────────────────────────────────────────────────────

export interface VersionInfo {
  version: string;
  environment: string;
  git_sha: string | null;
  git_sha_short: string | null;
  build_time: string | null;
  python_version: string | null;
  started_at: string;
  uptime_seconds: number;
}

export interface ReleaseRow {
  id: string;
  version: string;
  environment: string;
  git_sha: string | null;
  migration_revision: string | null;
  previous_version: string | null;
  is_rollback: boolean;
  deployed_at: string;
  verified: boolean;
  verification: Record<string, unknown>;
  notes: string | null;
}

export interface ReleaseInfo {
  current: VersionInfo;
  migration_current: string | null;
  migration_expected: string | null;
  migrations_at_head: boolean;
  latest_release: ReleaseRow | null;
  history: ReleaseRow[];
}

// ── Diagnostics ──────────────────────────────────────────────────────────────

export type HealthStatus = "healthy" | "degraded" | "unhealthy";
export type Severity = "critical" | "warning" | "info";

export interface DiagnosticCheck {
  name: string;
  group: string;
  severity: Severity;
  passed: boolean;
  detail: string;
}

export interface DiagnosticsReport {
  status: HealthStatus;
  checked_at: string;
  duration_ms: number;
  version: string;
  environment: string;
  passed_count: number;
  failed_count: number;
  critical_failures: string[];
  checks: DiagnosticCheck[];
}

export interface VerificationCheck {
  name: string;
  passed: boolean;
  detail: string;
}

export interface VerificationResult {
  passed: boolean;
  checked_at: string;
  version: string;
  environment?: string | null;
  checks: VerificationCheck[];
}

export interface RollbackVerification {
  passed: boolean;
  is_rollback: boolean;
  version: string;
  previous_version: string | null;
  checked_at: string;
  checks: VerificationCheck[];
}

// ── Status / metrics ─────────────────────────────────────────────────────────

export interface RouteLatency {
  method: string;
  path: string;
  count: number;
  avg_ms: number;
}

export interface MetricsSummary {
  uptime_seconds: number;
  total_requests: number;
  server_errors: number;
  client_errors: number;
  error_rate: number;
  avg_latency_ms: number;
  exceptions: Record<string, number>;
  events: Record<string, number>;
  slowest_routes: RouteLatency[];
}

export interface SystemStatus {
  status: HealthStatus;
  version: string;
  environment: string;
  checked_at: string;
  metrics: MetricsSummary;
  entities: Record<string, number>;
  active_users_24h: number;
  diagnostics: DiagnosticsReport;
}

// ── Backups ──────────────────────────────────────────────────────────────────

export interface BackupRow {
  id: string;
  scope: string;
  farm_id: string | null;
  label: string;
  trigger: string;
  status: string;
  checksum: string | null;
  size_bytes: number;
  record_counts: Record<string, number>;
  schema_version: string;
  app_version: string | null;
  started_at: string | null;
  completed_at: string | null;
  duration_ms: number | null;
  error: string | null;
  expires_at: string | null;
  created_at: string;
}

export interface BackupVerification {
  backup_id: string;
  valid: boolean;
  detail: string;
  expected_checksum: string | null;
  actual_checksum: string | null;
  size_bytes: number | null;
  record_counts: Record<string, number>;
}

export interface RestoreRunRow {
  id: string;
  backup_id: string;
  farm_id: string | null;
  dry_run: boolean;
  status: string;
  summary: Record<string, Record<string, number>>;
  checksum_verified: boolean;
  safety_backup_id: string | null;
  started_at: string | null;
  completed_at: string | null;
  duration_ms: number | null;
  error: string | null;
  created_at: string;
}

export interface RetentionResult {
  expired_removed: number;
  pruned_over_limit: number;
  retention_days: number;
  max_per_farm: number;
  swept_at: string;
}

// ── Imports / exports ────────────────────────────────────────────────────────

export interface ImportRowError {
  row: number;
  message: string;
}

export interface ImportJobRow {
  id: string;
  farm_id: string;
  entity: string;
  source_format: string;
  filename: string | null;
  dry_run: boolean;
  status: string;
  total_rows: number;
  valid_rows: number;
  imported_rows: number;
  failed_rows: number;
  errors: ImportRowError[];
  started_at: string | null;
  completed_at: string | null;
  duration_ms: number | null;
  error: string | null;
  created_at: string;
}

export interface ImportEntityInfo {
  entity: string;
  columns: string[];
  required: string[];
}

export interface ExportJobRow {
  id: string;
  farm_id: string;
  dataset: string;
  export_format: string;
  status: string;
  row_count: number;
  size_bytes: number;
  duration_ms: number | null;
  error: string | null;
  created_at: string;
}

export interface ExportDatasetInfo {
  dataset: string;
  formats: string[];
}

File: backend/app/schemas/production.py
"""
Greena — Production Readiness Schemas (Module 11).
"""

import uuid
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import Field

from app.schemas.base import AGRIOSSchema


# ── Version / release ─────────────────────────────────────────────────────────

class VersionInfo(AGRIOSSchema):
    version: str
    environment: str
    git_sha: Optional[str] = None
    git_sha_short: Optional[str] = None
    build_time: Optional[str] = None
    python_version: Optional[str] = None
    started_at: str
    uptime_seconds: int


class ReleaseRow(AGRIOSSchema):
    id: uuid.UUID
    version: str
    environment: str
    git_sha: Optional[str] = None
    migration_revision: Optional[str] = None
    previous_version: Optional[str] = None
    is_rollback: bool
    deployed_at: datetime
    verified: bool
    verification: dict = Field(default_factory=dict)
    notes: Optional[str] = None


class ReleaseInfo(AGRIOSSchema):
    """Everything the Release Information page renders."""

    current: VersionInfo
    migration_current: Optional[str] = None
    migration_expected: Optional[str] = None
    migrations_at_head: bool
    latest_release: Optional[ReleaseRow] = None
    history: list[ReleaseRow] = Field(default_factory=list)


# ── Diagnostics ───────────────────────────────────────────────────────────────

class DiagnosticCheck(AGRIOSSchema):
    name: str
    group: str
    severity: Literal["critical", "warning", "info"]
    passed: bool
    detail: str


class DiagnosticsReport(AGRIOSSchema):
    status: Literal["healthy", "degraded", "unhealthy"]
    checked_at: str
    duration_ms: int
    version: str
    environment: str
    passed_count: int
    failed_count: int
    critical_failures: list[str] = Field(default_factory=list)
    checks: list[DiagnosticCheck] = Field(default_factory=list)


class VerificationCheck(AGRIOSSchema):
    name: str
    passed: bool
    detail: str


class VerificationResult(AGRIOSSchema):
    passed: bool
    checked_at: str
    version: str
    environment: Optional[str] = None
    checks: list[VerificationCheck] = Field(default_factory=list)


class RollbackVerification(AGRIOSSchema):
    passed: bool
    is_rollback: bool
    version: str
    previous_version: Optional[str] = None
    checked_at: str
    checks: list[VerificationCheck] = Field(default_factory=list)


# ── System status / metrics ───────────────────────────────────────────────────

class RouteLatency(AGRIOSSchema):
    method: str
    path: str
    count: int
    avg_ms: float


class MetricsSummary(AGRIOSSchema):
    uptime_seconds: int
    total_requests: int
    server_errors: int
    client_errors: int
    error_rate: float
    avg_latency_ms: float
    exceptions: dict[str, int] = Field(default_factory=dict)
    events: dict[str, int] = Field(default_factory=dict)
    slowest_routes: list[RouteLatency] = Field(default_factory=list)


class SystemStatus(AGRIOSSchema):
    """The System Status page: health, metrics and entity counts in one call."""

    status: Literal["healthy", "degraded", "unhealthy"]
    version: str
    environment: str
    checked_at: str
    metrics: MetricsSummary
    entities: dict[str, int] = Field(default_factory=dict)
    active_users_24h: int = 0
    diagnostics: DiagnosticsReport


# ── Backups ───────────────────────────────────────────────────────────────────

class BackupRow(AGRIOSSchema):
    id: uuid.UUID
    scope: str
    farm_id: Optional[uuid.UUID] = None
    label: str
    trigger: str
    status: str
    checksum: Optional[str] = None
    size_bytes: int
    record_counts: dict[str, int] = Field(default_factory=dict)
    schema_version: str
    app_version: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    duration_ms: Optional[int] = None
    error: Optional[str] = None
    expires_at: Optional[datetime] = None
    created_at: datetime


class BackupCreateInput(AGRIOSSchema):
    label: Optional[str] = Field(default=None, max_length=200)
    retention_days: int = Field(default=30, ge=1, le=3650)


class BackupVerification(AGRIOSSchema):
    backup_id: str
    valid: bool
    detail: str
    expected_checksum: Optional[str] = None
    actual_checksum: Optional[str] = None
    size_bytes: Optional[int] = None
    record_counts: dict[str, int] = Field(default_factory=dict)


class RestoreInput(AGRIOSSchema):
    backup_id: uuid.UUID
    # Dry run by default: restoring is destructive, so it must be chosen twice.
    dry_run: bool = True
    overwrite: bool = False


class RestoreRunRow(AGRIOSSchema):
    id: uuid.UUID
    backup_id: uuid.UUID
    farm_id: Optional[uuid.UUID] = None
    dry_run: bool
    status: str
    summary: dict[str, Any] = Field(default_factory=dict)
    checksum_verified: bool
    safety_backup_id: Optional[uuid.UUID] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    duration_ms: Optional[int] = None
    error: Optional[str] = None
    created_at: datetime


class RetentionResult(AGRIOSSchema):
    expired_removed: int
    pruned_over_limit: int
    retention_days: int
    max_per_farm: int
    swept_at: str


# ── Imports ───────────────────────────────────────────────────────────────────

class ImportError_(AGRIOSSchema):
    row: int
    message: str


class ImportJobRow(AGRIOSSchema):
    id: uuid.UUID
    farm_id: uuid.UUID
    entity: str
    source_format: str
    filename: Optional[str] = None
    dry_run: bool
    status: str
    total_rows: int
    valid_rows: int
    imported_rows: int
    failed_rows: int
    errors: list[ImportError_] = Field(default_factory=list)
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    duration_ms: Optional[int] = None
    error: Optional[str] = None
    created_at: datetime


class ImportEntityInfo(AGRIOSSchema):
    entity: str
    columns: list[str]
    required: list[str]


# ── Exports ───────────────────────────────────────────────────────────────────

class ExportJobRow(AGRIOSSchema):
    id: uuid.UUID
    farm_id: uuid.UUID
    dataset: str
    export_format: str
    status: str
    row_count: int
    size_bytes: int
    duration_ms: Optional[int] = None
    error: Optional[str] = None
    created_at: datetime


class ExportDatasetInfo(AGRIOSSchema):
    dataset: str
    formats: list[str]
